libr/debug/p/debug_native.c

```diff
@@ -273,84 +285,107 @@ static int r_debug_native_wait (RDebug *dbg, int pid) {
-#if __linux__ && defined(PT_GETEVENTMSG)
-			// Handle PTRACE_EVENT_*
-			if (WIFSTOPPED (status) && WSTOPSIG (status) == SIGTRAP) {
-				ut32 pt_evt = status >> 16;
-				ut32 data;
-
-				switch (pt_evt) {
-				case 0:
-					// Normal trap?
-					break;
-
-				case PTRACE_EVENT_FORK:
-					if (dbg->trace_forks) {
-						if (ptrace (PTRACE_GETEVENTMSG, pid, 0, &data) == -1) {
-							r_sys_perror ("ptrace GETEVENTMSG");
-						} else {
-							eprintf ("PTRACE_EVENT_FORK new_pid=%d\n", data);
-							dbg->forked_pid = data;
-							// TODO: more handling here?
-						}
-					}
-					break;
-				case PTRACE_EVENT_EXIT:
-					if (ptrace (PTRACE_GETEVENTMSG, pid, 0, &data) == -1) {
-						r_sys_perror ("ptrace GETEVENTMSG");
-					} else {
-						eprintf ("PTRACE_EVENT_EXIT pid=%d, status=%d\n", pid, data);
-					}
-					break;
-				default:
-					eprintf ("Unknown PTRACE_EVENT encountered: %d\n", pt_evt);
-					break;
-				}
-			}
-#endif
+	//eprintf ("r_debug_native_wait: status=%d (0x%x) (return=%d)\n", status, status, ret);
 
-			r_debug_handle_signals (dbg);
+#ifdef WAIT_ON_ALL_CHILDREN
+	if (ret != pid) {
+		reason = R_DEBUG_REASON_NEW_PID;
+		eprintf ("switching to pid %d\n", ret);
+		r_debug_select(dbg, ret, ret);
+	}
+#endif // WAIT_ON_ALL_CHILDREN
 
-			if (WIFSTOPPED (status)) {
-				dbg->reason.signum = WSTOPSIG (status);
-				status = R_DEBUG_REASON_SIGNAL;
-			} else if (status == 0 || ret == -1) {
-				status = R_DEBUG_REASON_DEAD;
-			} else {
-				if (ret != pid)
-					status = R_DEBUG_REASON_NEW_PID;
-				else status = R_DEBUG_REASON_UNKNOWN;
-			}
+	// TODO: switch status and handle reasons here
+#if __linux__ && defined(PT_GETEVENTMSG)
+	reason = linux_ptrace_event (dbg, pid, status);
+#endif // __linux__
+
+	/* propagate errors */
+	if (reason == R_DEBUG_REASON_ERROR)
+		return reason;
+
+	/* we don't know what to do yet, let's try harder to figure it out. */
+	if (reason == R_DEBUG_REASON_UNKNOWN) {
+		if (WIFEXITED (status)) {
+			eprintf ("child exited with status %d\n", WEXITSTATUS(status));
+			reason = R_DEBUG_REASON_DEAD;
+		} else if (WIFSIGNALED (status)) {
+			eprintf ("child received signal %d\n", WTERMSIG(status));
+			reason = R_DEBUG_REASON_SIGNAL;
+		} else if (WIFSTOPPED (status)) {
+			eprintf ("child stopped with signal %d\n", WSTOPSIG (status));
+
+			/* this one might be good enough... */
+			dbg->reason.signum = WSTOPSIG (status);
+
+			/* the ptrace documentation says GETSIGINFO is only necessary for
+			 * differentiating the various stops.
+			 *
+			 * this might modify dbg->reason.signum
+			 */
+			if (!r_debug_handle_signals (dbg))
+				return R_DEBUG_REASON_ERROR;
+			reason = dbg->reason.type;
+		} else if (WIFCONTINUED (status)) {
+			eprintf ("child continued...\n");
+			reason = R_DEBUG_REASON_NONE;
+		} else if (status == 1) {
+			/* XXX(jjd): does this actually happen? */
+			eprintf ("EEK DEAD DEBUGEE!\n");
+			reason = R_DEBUG_REASON_DEAD;
+		} else if (status == 0) {
```

Comment:
oh right, i read it wrong. if it’s there is must be for something :P maybe git blame will light the commit that introduced this and we can identify the platform

> On 21 Jun 2016, at 00:56, Joshua J. Drake <notifications@github.com> wrote:
> 
> In libr/debug/p/debug_native.c <https://github.com/radare/radare2/pull/5185#discussion_r67781825>:
> 
> > +			/* the ptrace documentation says GETSIGINFO is only necessary for
> > +			 * differentiating the various stops.
> > +			 *
> > +			 * this might modify dbg->reason.signum
> > +			 */
> > +			if (!r_debug_handle_signals (dbg))
> > +				return R_DEBUG_REASON_ERROR;
> > +			reason = dbg->reason.type;
> > +		} else if (WIFCONTINUED (status)) {
> > +			eprintf ("child continued...\n");
> > +			reason = R_DEBUG_REASON_NONE;
> > +		} else if (status == 1) {
> > +			/* XXX(jjd): does this actually happen? */
> > +			eprintf ("EEK DEAD DEBUGEE!\n");
> > +			reason = R_DEBUG_REASON_DEAD;
> > +		} else if (status == 0) {
> status and reason are different things ;-)
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub <https://github.com/radare/radare2/pull/5185/files/e9879b72dc00a8bffb17ddb925ed028cf4920275#r67781825>, or mute the thread <https://github.com/notifications/unsubscribe/AA3-ljv3Dy2aA3po_MdFWbbVcnbijlbjks5qNxqcgaJpZM4I6ILO>.
>